### PR TITLE
feat: Org user table - bulk actions  

### PR DESCRIPTION
**Pull request by sean-brydon**
Loom: https://www.loom.com/share/5bd2f7c889e44ec48474a8776b31bc2e

Note: current implementation does not match the designs 100% more is to come with this feature. 
Figma: https://www.figma.com/file/ZJNAFHCAq6rnuGRgUFxjzQ/Organizations----New-features?type=design&node-id=4204-199174&mode=dev
How to test:

Create a org
create a bunch of teams
Add a bunch of users to the Org.
Edit accepted users
Use bulk actions

### CODE REVIEW

File: apps/web/tailwind.config.js [hunk 1]
```diff
@@ -3,4 +3,5 @@ const base = require("@calcom/config/tailwind-preset");
 module.exports = {
   ...base,
   content: [...base.content, "../../node_modules/@tremor/**/*.{js,ts,jsx,tsx}"],
+  plugins: [...base.plugins, require("tailwindcss-animate")],
```
Comment (sean-brydon) [inline apps/web/tailwind.config.js:6]: We only use this in web atm - not adding it to base components just yet.

File: apps/web/tsconfig.json
Comment (sean-brydon) [inline apps/web/tsconfig.json:1]: This happens every time i run yarn dev. @keithwillcode @zomars

File: packages/features/users/components/UserTable/BulkActions/DeleteBulkUsers.tsx [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+import type { User } from "../UserListTable";
+
+interface Props {
+  table: Table<User>;
+}
+
+export function DeleteBulkUsers({ table }: Props) {
+  const { t } = useLocale();
+  const selectedRows = table.getSelectedRowModel().flatRows.map((row) => row.original); // Get selected rows from table
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/BulkActions/DeleteBulkUsers.tsx:16]: Get selected row values from react-table

File: packages/features/users/components/UserTable/BulkActions/DeleteBulkUsers.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  const utils = trpc.useContext();
+  const deleteMutation = trpc.viewer.organizations.bulkDeleteUsers.useMutation({
+    onSuccess: () => {
+      utils.viewer.organizations.listMembers.invalidate();
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/BulkActions/DeleteBulkUsers.tsx:20]: Invalidate -> we may need to use optimistic UI when orgs get bigger. For now its not needed

File: packages/features/users/components/UserTable/BulkActions/TeamList.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+      showToast(
+        `${res.invitedTotalUsers} Users invited to ${Array.from(selectedValues).length} teams`,
+        "success"
+      );
+      // Optimistically update the data from query trpc cache listMembers
+      // We may need to set this data instread of invalidating. Will see how performance handles it
+      utils.viewer.organizations.listMembers.invalidate();
+
+      // Clear the selected values
+      setSelectedValues(new Set());
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/BulkActions/TeamList.tsx:47]: Clear the team selections once the users have been added to the team

File: packages/features/users/components/UserTable/BulkActions/TeamList.tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+      table.toggleAllRowsSelected(false);
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/BulkActions/TeamList.tsx:48]: Unselect all selected users in the Table

File: packages/features/users/components/UserTable/BulkActions/TeamList.tsx [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+              <CommandEmpty>No results found.</CommandEmpty>
+              <CommandGroup>
+                {teams &&
+                  teams.map((option) => {
+                    const isSelected = selectedValues.has(option.id);
+                    return (
+                      <CommandItem
+                        key={option.id}
+                        onSelect={() => {
+                          console.log("onSelect", isSelected);
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/BulkActions/TeamList.tsx:88]:
```suggestion
```

File: packages/features/users/components/UserTable/BulkActions/TeamList.tsx [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+                          }
+                        }}>
+                        <span>{option.name}</span>
+                        <div
+                          className={classNames(
+                            "border-subtle ml-auto flex h-4 w-4 items-center justify-center rounded-sm border",
+                            isSelected ? "text-emphasis" : "opacity-50 [&_svg]:invisible"
+                          )}>
+                          <Check className={classNames("h-4 w-4")} />
+                        </div>
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/BulkActions/TeamList.tsx:102]: We can't use a checkbox input component here as it hyjacks focus - making our own is easier

File: packages/features/users/components/UserTable/EditSheet/DisplayInfo.tsx [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+  label,
+  value,
+  asBadge,
+  isArray,
+  displayCopy,
+  displayCount,
+  badgeColor,
+}: DisplayInfoType<T>) {
+  const { copyToClipboard, isCopied } = useCopy();
+  const values = (isArray ? value : [value]) as string[];
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/EditSheet/DisplayInfo.tsx:28]: Always make it an array internally so we can just loop.

File: packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+import { useEditMode } from "./store";
+
+const editSchema = z.object({
+  name: z.string(),
+  email: z.string().email(),
+  bio: z.string(),
+  role: z.enum(["ADMIN", "MEMBER"]),
+  timeZone: z.string(),
+  // schedules: z.array(z.string()),
+  // teams: z.array(z.string()),
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx:40]: These are posible option to come in the sheet.

File: packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+    onSettled: () => {
+      /**
+       * /We need to do this as the submit button lives out side
+       *  the form for some complicated reason so we can't relay on mutationState
+       */
+      setMutationLoading(false);
+    },
+  });
+
+  const watchTimezone = form.watch("timeZone");
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx:88]: Select either needed a controller or using watch - this was easier

File: packages/features/users/components/UserTable/EditSheet/SheetFooterControls.tsx [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        color="secondary"
+        type="button"
+        className="justify-center md:w-1/5"
+        onClick={() => {
+          setEditMode(false);
+        }}>
+        {t("cancel")}
+      </Button>
+
+      <Button type="submit" className="w-full justify-center" form="edit-user-form" loading={isLoading}>
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/EditSheet/SheetFooterControls.tsx:24]: Has a form="..."

as it points to a form that its not inside of. 

easiest way to do this. We do something like this in event-type-settings

File: packages/features/users/components/UserTable/EditSheet/store.ts
Comment (sean-brydon) [inline packages/features/users/components/UserTable/EditSheet/store.ts:1]: Global state used for this component - easier than using context imho

File: packages/features/users/components/UserTable/UserListTable.tsx [hunk 12]
```diff
@@ -42,11 +45,17 @@ export type State = {
...
 
 export type Action =
   | {
-      type: "SET_CHANGE_MEMBER_ROLE_ID" | "SET_DELETE_ID" | "SET_IMPERSONATE_ID" | "INVITE_MEMBER";
+      type:
+        | "SET_CHANGE_MEMBER_ROLE_ID"
+        | "SET_DELETE_ID"
+        | "SET_IMPERSONATE_ID"
+        | "INVITE_MEMBER"
+        | "EDIT_USER_SHEET";
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/UserListTable.tsx:58]: Adds another action for the user sheet

File: packages/features/users/components/UserTable/UserListTable.tsx [hunk 13]
```diff
@@ -121,25 +136,25 @@ export function UserListTable() {
...
+        ),
+        cell: ({ row }) => (
+          <Checkbox
+            checked={row.getIsSelected()}
+            onCheckedChange={(value) => row.toggleSelected(!!value)}
+            aria-label="Select row"
+            className="translate-y-[2px]"
+          />
+        ),
+      },
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/UserListTable.tsx:157]: Re-enable select as this PR introduces the functionality

File: packages/features/users/components/UserTable/UserListTable.tsx [hunk 14]
```diff
@@ -269,18 +284,12 @@ export function UserListTable() {
         searchKey="member"
         selectionOptions={[
           {
-            label: "Add To Team",
-            onClick: () => {
-              console.log("Add To Team");
-            },
-            icon: Users,
+            type: "render",
+            render: (table) => <TeamListBulkAction table={table} />,
```
Comment (sean-brydon) [inline packages/features/users/components/UserTable/UserListTable.tsx:288]: Add the option to render components in the bar instead of just onclick handlers.

Easily allows us to use dialogTrigger to get the correct position of the popover to open. 

I tried without and ended up overcomplicating the calculation with refs :/

File: packages/lib/hooks/useCopy.ts
Comment (sean-brydon) [inline packages/lib/hooks/useCopy.ts:1]: quick useCopy hook that doesnt throw errors on the server and works when it hits the client

File: packages/trpc/server/routers/viewer/organizations/addBulkTeams.handler.ts [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+    where: {
+      user: {
+        organizationId: currentUser.organizationId,
+        id: {
+          in: input.userIds,
+        },
+      },
+      accepted: true,
+    },
+    distinct: ["userId"],
```
Comment (sean-brydon) [inline packages/trpc/server/routers/viewer/organizations/addBulkTeams.handler.ts:38]: We distinct as the user could have a membership that is not accepted

File: packages/ui/components/data-table/DataTableSelectionBar.tsx [hunk 16]
```diff
@@ -1,16 +1,25 @@
...
+  | {
+      type: "action";
+      label: string;
+      onClick: () => void;
+      icon?: SVGComponent;
+    }
+  | {
+      type: "render";
+      render: (table: Table<TData>) => React.ReactNode;
+    };
```
Comment (sean-brydon) [inline packages/ui/components/data-table/DataTableSelectionBar.tsx:18]: Support the ability to render custom components instead of a predefined action.

File: packages/ui/components/data-table/DataTableSelectionBar.tsx [hunk 17]
```diff
@@ -21,10 +30,16 @@ export function DataTableSelectionBar<TData>({ table, actions }: DataTableSelect
...
-          {action.label}
-        </Button>
+      {actions?.map((action, index) => (
+        <Fragment key={index}>
+          {action.type === "action" ? (
+            <Button aria-label={action.label} onClick={action.onClick} StartIcon={action.icon}>
+              {action.label}
+            </Button>
+          ) : action.type === "render" ? (
+            action.render(table)
```
Comment (sean-brydon) [inline packages/ui/components/data-table/DataTableSelectionBar.tsx:40]: Pass table as a callback -> Pass state out of the DataTableComponent and into the user sheet for example

File: apps/web/package.json [hunk 18]
```diff
@@ -167,6 +167,7 @@
     "msw": "^0.42.3",
     "postcss": "^8.4.18",
     "tailwindcss": "^3.3.1",
+    "tailwindcss-animate": "^1.0.6",
```
Comment (leog) [inline apps/web/package.json:170]: Wondering if we measured the impact of adding such plugins...

File: apps/web/tsconfig.json [hunk 19]
```diff
@@ -3,12 +3,28 @@
...
+        "server/*"
+      ],
+      "@prisma/client/*": [
+        "@calcom/prisma/client/*"
+      ]
+    },
+    "plugins": [
+      {
+        "name": "next"
+      }
```
Comment (leog) [inline apps/web/tsconfig.json:25]: Why was this necessary?

File: packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+              <SheetFooterControls />
+            </SheetFooter>
+          </div>
+        ) : (
+          <Loader />
+        )}
+      </SheetContent>
+    </Sheet>
+  );
+}
```
Comment (leog) [inline packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx:249]: Quite a big file, not sure if it makes sense to separate in different files

File: packages/trpc/server/routers/viewer/organizations/addBulkTeams.handler.ts [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+  input: TAddBulkTeams;
+};
+
+export async function addBulkTeamsHandler({ ctx, input }: AddBulkTeamsHandler) {
+  const currentUser = ctx.user;
+
+  if (!currentUser.organizationId) throw new TRPCError({ code: "UNAUTHORIZED" });
+
+  // check if user is admin of organization
+  if (!(await isOrganisationAdmin(currentUser?.id, currentUser.organizationId)))
```
Comment (leog) [inline packages/trpc/server/routers/viewer/organizations/addBulkTeams.handler.ts:24]: Might be worth creating a trpc procedure to check this on the router

File: yarn.lock [hunk 22]
```diff
@@ -131,25 +131,6 @@ __metadata:
   languageName: node
   linkType: hard
 
-"@auth/core@npm:^0.1.4":
```
Comment (leog) [inline yarn.lock:134]: Not sure if we want to let update the entire `yarn.lock` when adding a new dependency, maybe just pick up the changes to introduce the dependency and leave the rest out for its own updating process?

**Review by leog — COMMENTED** [review]
Code looks good, leaving some comments.

File: apps/web/package.json (hunk 18 above)
Comment (sean-brydon) [reply apps/web/package.json:170]: Its just a CSS addition - will just increase css bundle if anything. Very miniscule

File: apps/web/tsconfig.json (hunk 19 above)
Comment (sean-brydon) [reply apps/web/tsconfig.json:25]: As mentioned in the comment - this is added whenever i run yarn dev on main. 

next says its adding a plugin due to reccomendations.

File: packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx (hunk 20 above)
Comment (sean-brydon) [reply packages/features/users/components/UserTable/EditSheet/EditUserSheet.tsx:249]: Will do this in the morning

File: packages/trpc/server/routers/viewer/organizations/addBulkTeams.handler.ts (hunk 21 above)
Comment (sean-brydon) [reply packages/trpc/server/routers/viewer/organizations/addBulkTeams.handler.ts:24]: Maybe - this is a NIT tho and can be refactored along with the rest of org routes that require admin.

File: apps/web/package.json (hunk 18 above)
Comment (sean-brydon) [reply apps/web/package.json:170]: And since its tailwind - JIT just compiles the classes down to only the ones we use. This might add 2 css classes

**Comment by hariombalhara** [comment]
nit: Enterprise feature banner on local is causing the popover to be partially hidden
image
Note a big problem because you can still scroll and this would happen only on local.

File: packages/features/users/components/UserTable/UserListTable.tsx (hunk 14 above)
Comment (hariombalhara) [reply packages/features/users/components/UserTable/UserListTable.tsx:288]: image

Team that the users are already a member of are shown in Bulk Add to Team. This is the case even when all the selected users were part of that team. I think we should disable that team or not show at all.

But when some of them are a part of it and some of them are not, it currently says that all the users have been invited when infact only those that are not part of the team are added to the team. Maybe we can handle it by showing something like '2 members already a part and 3 were added'

File: packages/features/users/components/UserTable/BulkActions/TeamList.tsx [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+                      </CommandItem>
+                    );
+                  })}
+              </CommandGroup>
+              <>
+                <CommandSeparator />
+                <CommandGroup>
+                  <div className="mb-1.5 flex w-full">
+                    <Button
```
Comment (hariombalhara) [inline packages/features/users/components/UserTable/BulkActions/TeamList.tsx:111]: This button is not visible when we do a search for team.
image

File: packages/features/users/components/UserTable/BulkActions/DeleteBulkUsers.tsx (hunk 2 above)
Comment (hariombalhara) [reply packages/features/users/components/UserTable/BulkActions/DeleteBulkUsers.tsx:16]: nit: This component needs just the users. Maybe we can pass just that prop to DeleteBulkUsers
Screenshot 2023-08-02 at 5 15 39 PM

File: packages/features/users/components/UserTable/BulkActions/TeamList.tsx [hunk 24]
```diff
@@ -0,0 +1,127 @@
...
+
+import type { User } from "../UserListTable";
+
+interface Props {
+  table: Table<User>;
+}
+
+export function TeamListBulkAction({ table }: Props) {
+  const { data: teams } = trpc.viewer.organizations.getTeams.useQuery();
+  const [selectedValues, setSelectedValues] = useState<Set<number>>(new Set());
```
Comment (hariombalhara) [inline packages/features/users/components/UserTable/BulkActions/TeamList.tsx:30]: Love the use of Set here. I got to teach myself to remember use it.

**Review by leog — DISMISSED** [review]
Thanks for the replies on my comments. LGTM

**Comment by PeerRich** [comment]
how does this look mobile and tablet @sean-brydon ?

**Comment by sean-brydon** [comment]
> how does this look mobile and tablet @sean-brydon ?
CleanShot 2023-08-14 at 09 24 36@2x
Like this - overflow-x becomes scrollable

File: packages/trpc/server/routers/viewer/organizations/_router.tsx
Comment (zomars) [inline packages/trpc/server/routers/viewer/organizations/_router.tsx:1]: I hate boilerplate

File: packages/trpc/server/routers/viewer/organizations/adminVerify.handler.ts [hunk 25]
```diff
@@ -114,3 +114,5 @@ export const adminVerifyHandler = async ({ input }: AdminVerifyOptions) => {
     message: `Verified Organization - Auto accepted all members ending in ${acceptedEmailDomain}`,
   };
 };
+
+export default adminVerifyHandler;
```
Comment (zomars) [inline packages/trpc/server/routers/viewer/organizations/adminVerify.handler.ts:118]: Had to add default exports in order to use importHandler helper function.

File: packages/trpc/server/routers/viewer/organizations/updateUser.handler.ts [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+        userId_teamId: {
+          userId: input.userId,
+          teamId: organizationId,
+        },
+      },
+      data: {
+        role: input.role,
+      },
+    }),
+  ]);
```
Comment (zomars) [inline packages/trpc/server/routers/viewer/organizations/updateUser.handler.ts:60]: @sean-brydon just curious, are there any pros/cons of doing this over a transaction?

```suggestion
  await prisma.user.update({
    where: {
      id: input.userId,
    },
    data: {
      bio: input.bio,
      email: input.email,
      name: input.name,
      timeZone: input.timeZone,
      organization: {
        update: {
          members: {
            update: {
              where: {
                userId_teamId: {
                  userId: input.userId,
                  teamId: organizationId,
                },
              },
              data: {
                role: input.role,
              },
            },
          },
        },
      },
    },
  });
```

**Review by zomars — APPROVED** [review]
LGTM
